File: 백준/Silver/9095. 1， 2， 3 더하기/1， 2， 3 더하기.py
#1,2,3 더하기
#4 -> 1,1,1,1(1) 1,1,2(3) 1,3(2)  2+2 (7)
# dp[3] + dp[2] + dp[1]

#1 -> 1 (1) 
#2 -> 1,1 ,2  (2)
#3 -> 1,1,1  3,1 ,2 2, 1 3  (4)
#4 -> 3,1 1,3 2,2 -> 
#종류 *1 -> 1 종류 n 개수 k -> 종류-1 * k

n = int(input())
nums = [int(input()) for _ in range(n)] 

for N in nums:
    dp = [0] * (N + 1)

    for i in range(1, N + 1):
        if i == 1:
            dp[i] = 1
        elif i == 2:
            dp[i] = 2
        elif i == 3:
            dp[i] = 4
        else:
            dp[i] = dp[i - 1] + dp[i - 2] + dp[i - 3]

    print(dp[N])

    

# 

File: 백준/Silver/9095. 1， 2， 3 더하기/1， 2， 3 더하기.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(filePath).toString().trim();

function solution(input) {
  const parseInput = input.split(`\n`);
  const N = Number(parseInput[0]);
  const nums = parseInput.slice(1).map(Number);
  const dp = new Array(Math.max(...nums)).fill(0);

  dp[0] = 1;
  dp[1] = 2;
  dp[2] = 4;

  for (let i = 3; i < dp.length; i++) {
    dp[i] = dp[i - 1] + dp[i - 2] + dp[i - 3];
  }

  const result = [];

  for (let i = 0; i < nums.length; i++) {
    result.push(dp[nums[i] - 1]);
  }

  return result.join(`\n`);
}

console.log(solution(input));
